### ISSUE TITLE
KeyError(1) while parsing output from applying flake8

### ISSUE THREAD

**Issue by psychemedia** [nbQA-dev/nbQA#588]
Hi

Running `nbqA flake8` over notebooks in a set of directories, I encountered a few errors of the form:

```
KeyError(1) while parsing output from applying flake8 to ...
```

I will try to dig into the notebooks to see what the problems appear to be and raise a separate issues as appropriate, but more generally it would be useful if `nbQA` could trap these errors rather than erroring and then publishing a notice or error file capturing the errors.

In passing, would it be worth adding to docs a simple example recipe for excluding multiple files with different filename patterns for folk who aren't overly familiar with regex? eg I am using a pattern of the form `nbqa flake8 --nbqa-exclude "(.*02.2.*)|(.*07.1.*)|(.*07.2.*)"  Part\ *\ Notebooks/*.ipynb` with `|` separated bracketed expressions.

**Comment by MarcoGorelli** [comment]
> no code cells! 

I can't reproduce this, I just tried running `nbqa flake8 tests/data/empty_notebook.ipynb` and it didn't report anything. Do you remember which command you ran? Could you show the output of `flake8 --version` please? Thanks 👍

**Comment by psychemedia** [comment]
Here's a [file](https://pastebin.com/868yBUVB) that fails for me (will expire in a week) with a simple `nbqa flake FILENAME` command.

**Comment by MarcoGorelli** [comment]
Still can't reproduce:
```
% nbqa --version        
nbqa 0.8.0
% flake8 --version        
3.8.4 (mccabe: 0.6.1, pycodestyle: 2.6.0, pyflakes: 2.2.0) CPython 3.8.5 on Darwin
% nbqa flake8 t.ipynb     
t.ipynb:cell_2:1:121: E501 line too long (202 > 120 characters)
t.ipynb:cell_11:1:1: E402 module level import not at top of file
```

Does `flake8 --version` not show anything between the brackets for you?

**Comment by psychemedia** [comment]
In situ:

```
nbqa --version        
nbqa 0.8.0

flake8 --version    
3.8.4 () CPython 3.9.4 on Darwin

nbqa flake8 Part\ *7*/*.ipynb
KeyError(1) while parsing output from applying flake8 to Part 07 Notebooks/07.1 Spreadsheet basics.ipynb
Please report a bug at https://github.com/nbQA-dev/nbQA/issues 
```

I'm running over may 100 notebooks and that error report on maybe 10 of them, five of which have no code cells. I will try to investigate what is throwing the issue in erroring notebooks.

**Comment by MarcoGorelli** [comment]
Does the error happen even if you only run on `Part 07 Notebooks/07.1 Spreadsheet basics.ipynb` alone, or only if you run it on `Part\ *7*/*.ipynb`?

**Comment by psychemedia** [comment]
The regex works fine just as a shortcut into a set of files (which makes it easy for me to run checks over lots of notebooks); the issue definitely that file...

```
nbqa flake8 Part\ 07\ Notebooks/07.1\ Spreadsheet\ basics.ipynb 
KeyError(1) while parsing output from applying flake8 to Part 07 Notebooks/07.1 Spreadsheet basics.ipynb
Please report a bug at https://github.com/nbQA-dev/nbQA/issues 
```

(Apols, not had a chance to dig into the detail debug myself; my immediate concerns are fixing the notebooks, not the debug tools helping me identify issues in the notebooks...)

**Comment by MarcoGorelli** [comment]
No worries about debugging it, but is `Part\ 07\ Notebooks/07.1\ Spreadsheet\ basics.ipynb ` the file which you posted above in pastebin? If so, I'm really confused about why it would fail for you but not for me. If not, any chance you could post that exact notebook somewhere and I'll take a look?

**Comment by psychemedia** [comment]
Yep, same file...

https://gist.github.com/psychemedia/2e05c7a4aac8002bf9a26c7d97920a9f

Could there be any side effects from other flake8 packages (either present or absent?). Possibly unusual things in the notebook include jupytext markup (I also run dev versions of jupytext in pretty much all my environments).

**Comment by psychemedia** [comment]
Other things in various notebooks - code blocks in markdown cells, including language qualified fenced codeblocks in markdown cells.

**Comment by MarcoGorelli** [comment]
> Could there be any side effects from other flake8 packages (either present or absent?)

Yes, that's what I'm thinking...do you remember if you have any `flake8` extensions installed? Perhaps could you show the output of `pip freeze`? @s-weigand any ideas?

Thanks for being so helpful here with debugging by the way, really appreciate it! 🙏

**Comment by psychemedia** [comment]
flake referenced packages in filename... (thanks for being so persistent with this, debuggong a possibly problem w/ my setyup....:-)

```
pip3 freeze |grep .*flake.*
flake8==3.8.4
flake8-bugbear==20.11.1
flake8-builtins==1.5.3
flake8-docstrings==1.5.0
flake8-nb @ file:///Users/tonyhirst/Documents/GitHub/flake8-nb
flake8-polyfill==1.0.2
nb-report-injecter @ file:///Users/tonyhirst/Documents/GitHub/nbqa-inject-flake8-report
pyflakes==2.2.0
```
I'll upgrade all those now to see if it helps...

**Comment by psychemedia** [comment]
Okay.. that errored; just uninstalled everything, including `nbqa` and then installed latest `nbqa` from pip again; that gave a missing `flake8` dependency; installed that, tried again, missing `flake8-polyfill` dependency, tried again and it passed.

So from this I learn: `nbqa` has undeclared dependency on `flake8` and `flake8-polyfill`.

If I now try to add things back in:

- `flake8-builtins` continues to work; 
- `flake8-docstrings` fails with `KeyError(1) while parsing output from applying flake8`
- uninstall `flake8-docstrings`, install `flake8-bugbear`, continues to work.

So `flake8-docstrings-1.6.0` is messing things up.

**Comment by MarcoGorelli** [comment]
Nice! Thanks @psychemedia , now I can reproduce the error!

> So from this I learn: nbqa has undeclared dependency on flake8 and flake8-polyfill.

Just to be clear, `nbqa` just requires you to have installed whatever tool you want to run. E.g. to run `nbqa flake8`, you need to have installed `flake8`, to run `nbqa pyupgrade`, you need to have installed `pyupgrade`, etc.

**Comment by MarcoGorelli** [comment]
@all-contributors please add @psychemedia for bugs

**Comment by MarcoGorelli** [comment]
Hey @psychemedia - I believe this should be fixed in version 0.8.1, would really appreciate it if you could install it and check (`pip install nbqa==0.8.1`)

**Comment by psychemedia** [comment]
Yep, that's free of errors for me on that notebook now. Thanks :-)
